nntrainer/include/databuffer_v2.h

```diff
@@ -0,0 +1,325 @@
+// SPDX-License-Identifier: Apache-2.0-only
+/**
+ * Copyright (C) 2020 Parichay Kapoor <pk.kapoor@samsung.com>
+ *
+ * @file	databuffer_v2.h
+ * @date	3 September 2020
+ * @see		https://github.com/nnstreamer/nntrainer
+ * @author	Parichay Kapoor <pk.kapoor@samsung.com>
+ * @bug		No known bugs except for NYI items
+ * @brief	This is databuffer class for Neural Network
+ *
+ * @todo TODO: Support multi files for dataset with files
+ * @todo TODO: Support multi threads with more than 1 thread and use thread
+ * pooling
+ * @todo TODO: Support label size to be 0 for inference based scenarios
+ * @todo TODO: rename data buffer to dataset
+ * @todo TODO: manage with just 1 buffer
+ * @todo TODO: consider different data structure for buffer when using shuffle
+ */
+
+#ifndef __DATABUFFER_V2_H__
+#define __DATABUFFER_V2_H__
+#ifdef __cplusplus
+
+#include <condition_variable>
+#include <list>
+#include <memory>
+#include <mutex>
+#include <string>
+#include <thread>
+#include <tuple>
+#include <vector>
+
+#include <nntrainer-api-common.h>
+#include <tensor.h>
+
+namespace nntrainer {
+
+/**
+ * @brief   States the collect thread running in the background
+ */
+enum class ThreadStates {
+  THREAD_NULL,            /**< not yet initialized */
+  THREAD_READY,           /**< initialized but not yet started */
+  THREAD_RUNNING,         /**< started and running */
+  THREAD_REQUEST_TO_STOP, /**< main thread has request to stop, background
+                             thread is in the process of being stopped */
+  THREAD_STOPPED, /**< background thread has stopped but not yet cleaned up */
+  THREAD_EPOCH_FINISHED, /**< background threads have finished the epoch */
+  THREAD_ERROR           /**< an error has occurred in background threads */
```

Comment:
I think it'd be better if we separate this to threadResult? because those state are not mutually exclusive.

eg. A thread can be  `thread_request_to_stop` and `thread_error`

Reply:
Yes, they are not mutually exclusive. However, they have priorities in a way. 
Once thread state is THREAD_ERROR, no more state change.
It cant be in both states.